File: Terrafirma/SteamConfig.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

namespace Terrafirma
{
    class SteamConfig
    {
        class Element
        {
            public Dictionary<string, Element> Children { get; set; }
            public string Name { get; set; }
            public string Value;

            private static Regex keyRegex = new Regex("\"[^\"]*\"");

            public Element(Queue<string> lines)
            {
                Children = new Dictionary<string, Element>();
                string line = lines.Dequeue();

                MatchCollection matches = keyRegex.Matches(line);
                if (matches.Count == 0) //corrupt
                    return;
                Name = matches[0].Value.Trim('\"');
                if (matches.Count > 1) //value is a string
                    Value = matches[1].Value.Trim('\"').Replace(@"\\","\\");
                line = lines.Peek();
                if (line.Contains('{'))
                {
                    lines.Dequeue();
                    while (true)
                    {
                        line = lines.Peek();
                        if (line.Contains('}'))
                        {
                            lines.Dequeue();
                            return;
                        }
                        Element e = new Element(lines);
                        Children.Add(e.Name, e);
                    }
                }
            }
            public string Find(string path)
            {
                string[] kv = path.Split(new char[] { '/' }, 2);
                if (kv.Length == 1)
                    return Children[kv[0]].Value;
                return Children[kv[0]].Find(kv[1]);
            }
        }
        Element root = null;
        public SteamConfig()
        {
            Microsoft.Win32.RegistryKey key;
            key = Microsoft.Win32.Registry.CurrentUser.OpenSubKey(@"Software\\Valve\\Steam");
            if (key != null)
            {
                string path = key.GetValue("SteamPath") as string;
                path = Path.Combine(path, "config");
                path = Path.Combine(path, "config.vdf");
                if (File.Exists(path))
                    parse(path);
            }
        }
        public bool Ready { get { return root != null; } }

        public string Get(string path)
        {
            return root.Find(path);
        }

        private void parse(string path)
        {
            TextReader read = new StreamReader(path);
            Queue<string> lines = new Queue<string>();
            String line = read.ReadLine();
            while (line != null)
            {
                lines.Enqueue(line);
                line = read.ReadLine();
            }
            root = new Element(lines);
        }

    }
}

File: steamconfig.cpp
/**
 * @Copyright 2015 seancode
 *
 * Handles parsing of the steam config files
 */

#include "./steamconfig.h"
#include <QSettings>
#include <QTextStream>
#include <QRegularExpression>
#include <QRegularExpressionMatch>
#include <QRegularExpressionMatchIterator>
#include <QStandardPaths>
#include <QDir>

SteamConfig::SteamConfig() {
  root = nullptr;
  QSettings settings("HKEY_CURRENT_USER\\Software\\Valve\\Steam",
                     QSettings::NativeFormat);
  QString path = settings.value("SteamPath").toString();
  if (path.isEmpty()) {
    path =  QStandardPaths::standardLocations(QStandardPaths::GenericDataLocation)
        .first();
    path += QDir::toNativeSeparators("/Steam");
  }
  path += QDir::toNativeSeparators("/config/config.vdf");
  QFile file(path);
  if (file.exists())
    parse(path);
}

QString SteamConfig::operator[](const QString &path) const {
  if (root == nullptr)
    return QString();
  return root->find(path);
}

void SteamConfig::parse(const QString &filename) {
  QFile file(filename);

  if (file.open(QIODevice::ReadOnly)) {
    QList<QString> strings;
    QTextStream in(&file);
    while (!in.atEnd())
      strings.append(in.readLine());
    file.close();
    root = new Element(&strings);
  }
}

SteamConfig::Element::Element() = default;

SteamConfig::Element::Element(QList<QString> *lines) {
  QString line;
  QRegularExpression re("\"([^\"]*)\"");
  QRegularExpressionMatchIterator i;
  while (lines->length() > 0) {
    line = lines->front();
    lines->pop_front();
    i = re.globalMatch(line);
    if (i.hasNext())
      break;
  }
  if (!lines->length())  // corrupt
    return;
  QRegularExpressionMatch match = i.next();
  name = match.captured(1).toLower();
  if (i.hasNext()) {  // value is a string
    match = i.next();
    value = match.captured(1);
    value.replace("\\\\", "\\");
  }
  line = lines->front();
  if (line.contains("{")) {
    lines->pop_front();
    while (true) {
      line = lines->front();
      if (line.contains("}")) {  // empty
        lines->pop_front();
        return;
      }
      Element e(lines);
      children[e.name] = e;
    }
  }
}

QString SteamConfig::Element::find(const QString &path) {
  int ofs = path.indexOf("/");
  if (ofs == -1)
    return children[path].value;
  return children[path.left(ofs)].find(path.mid(ofs + 1));
}
